src/PythonClient/tests/quixstreams/test_dataframes/test_kafka/test_consumer.py

```diff
@@ -0,0 +1,956 @@
+import asyncio
+import time
+import uuid
+from concurrent.futures import Future
+from typing import List
+
+import confluent_kafka
+import pytest
+from confluent_kafka.error import KafkaError
+from confluent_kafka import TopicPartition
+
+
+DEFAULT_TIMEOUT = 15.0
+
+
+async def wait_for_future(fut: Future, timeout: float = DEFAULT_TIMEOUT):
+    wrapped_fut = asyncio.wrap_future(fut)
+    return await asyncio.wait_for(wrapped_fut, timeout=timeout)
+
+
+async def test_consumer_start_close(
+    consumer,
+):
+    async with consumer:
+        metadata = await consumer.list_topics()
+        assert metadata
+        broker_meta = metadata.brokers[0]
+        broker_address = f"{broker_meta.host}:{broker_meta.port}"
+        assert broker_address
+
+    assert not consumer.running
+
+
+async def test_consumer_close_inner_consumer(
+    consumer,
+):
+    async with consumer:
+        metadata = await consumer.list_topics()
+        assert metadata
+        broker_meta = metadata.brokers[0]
+        broker_address = f"{broker_meta.host}:{broker_meta.port}"
+        assert broker_address
+        await consumer.close()
+        with pytest.raises(RuntimeError) as raised:
+            await consumer.poll(timeout=0)
+    assert str(raised.value) == "Consumer closed"
+
+
+class TestAsyncConsumerSubscribe:
+    async def test_consumer_subscribe_topic_exists(
+        self,
+        consumer,
+        topic_factory,
+    ):
+        topic_name, num_partitions = await topic_factory()
+
+        async with consumer:
+            await consumer.subscribe(topics=[topic_name])
+            msg = await consumer.poll(timeout=1)
+            assert msg is None
```

Comment:
One thing to consider checking as well: if you do a consumer.poll with a topic that has messages on it (and of course, a consumer that is not on the latest offset), make sure that polling doesn't accidentally consume a message.

I don't fully remember the circumstances, but I think I was polling to get metadata updates and it was possible to get a message if it was ready, which if you do not handle correctly, you will miss that message.